src/object/cli_coll.c

```diff
@@ -57,20 +221,74 @@ obj_coll_oper_args_init(struct coll_oper_args *coa, struct dc_object *obj, bool
 void
 obj_coll_oper_args_fini(struct coll_oper_args *coa)
 {
-	daos_coll_target_cleanup(coa->coa_dcts,
-				 coa->coa_dct_nr < 0 ? coa->coa_dct_cap : coa->coa_dct_nr);
-	coa->coa_dcts = NULL;
+	if (coa->coa_sparse) {
+		if (coa->coa_tree != NULL) {
+			if (daos_handle_is_valid(coa->coa_tree->cst_tree_hdl))
+				dbtree_destroy(coa->coa_tree->cst_tree_hdl, NULL);
+			D_FREE(coa->coa_tree);
+		}
+	} else {
+		daos_coll_target_cleanup(coa->coa_dcts,
+					 coa->coa_dct_nr < 0 ? coa->coa_dct_cap : coa->coa_dct_nr);
+		coa->coa_dcts = NULL;
+	}
 	coa->coa_dct_cap = 0;
 	coa->coa_dct_nr = 0;
 }
 
-int
-obj_coll_oper_args_collapse(struct coll_oper_args *coa, uint32_t *size)
+static int
+obj_coll_tree_cb(daos_handle_t ih, d_iov_t *key, d_iov_t *val, void *arg)
+{
+	struct coll_oper_args	*coa = arg;
+	struct daos_coll_target	*dct = val->iov_buf;
+
+	D_ASSERTF(coa->coa_dct_nr < coa->coa_dct_cap,
+		  "Too short pre-allcoated dct_array: %u vs %u\n",
+		  coa->coa_dct_nr, coa->coa_dct_cap);
+
+	memcpy(&coa->coa_dcts[coa->coa_dct_nr++], dct, sizeof(*dct));
+
+	/* The following members have been migrated into coa->coa_dcts. */
+	dct->dct_bitmap = NULL;
+	dct->dct_shards = NULL;
+	dct->dct_tgt_ids = NULL;
+
+	return 0;
+}
+
+static int
+obj_coll_collapse_tree(struct coll_oper_args *coa, uint32_t *size)
+{
+	struct coll_sparse_targets	*tree = coa->coa_tree;
+	int				 rc = 0;
+
+	if (unlikely(coa->coa_dct_cap == 0))
+		D_GOTO(out, rc = 1);
```

Comment:
this func's caller obj_coll_oper_args_collapse() does not differenciate rc 1 or rc 0, why here return 1 rather 0?

Reply:
Returning 1 will make the caller to "goto out" with positive number, that will not fail the operation.